Add Opentherm Gateway current and setpoint precision

Martidjen opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Dividing temperature precision into precision for current temperature and precision for setpoint temperature


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests


## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/16904

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Martidjen committed:
First attempt for otgw improvements

Devided precision into read and set precision.
Added option for setpoint override mode, can be set to temporary or constant (some thermostats work better with constant mode)

Martidjen committed:
Attempt 2 - bugfix and translations

Fixing some issues found while testing and adding translations for new options

Commit:
Updating tests for new options

Martidjen committed:
Fixing some issues after running local tests

Fixing some issues after running local tests

homeassistant commented:
Hello @Martidjen,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`51309cf19a255e93df03785381472c41f8be42f2`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;core&#x2F;pull&#x2F;47484&#x2F;commits&#x2F;51309cf19a255e93df03785381472c41f8be42f2) - This commit has something that looks like an email address (martijn.wals@boonedam.com). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

Martidjen committed:
PR check fixes

Setting rights en sorting order to satisfy the checks in the pull request

mvn23 commented:
Translations should be handled separately, see [here](https://developers.home-assistant.io/docs/translations).
Please remove them from the PR.

Martidjen committed:
Reverting translations

Reverting changes to translations after comment on PR

Martidjen commented:
Reverted my changes to the translations, thank you for replying

Martidjen committed:
Corrected some more translations

Found some more changes to be corrected

MartinHjelmare commented on homeassistant/components/opentherm_gw/__init__.py:
```diff
@@ -59,9 +63,15 @@
+        vol.Optional(CONF_READ_PRECISION): vol.In(
             [PRECISION_TENTHS, PRECISION_HALVES, PRECISION_WHOLE]
         ),
+        vol.Optional(CONF_SET_PRECISION): vol.In(
```
We no longer accept additions or changes to the yaml config of integrations. See: https://github.com/home-assistant/architecture/blob/master/adr/0010-integration-configuration.md

This integration needs to be refactored to use a config flow and config entries.

https://developers.home-assistant.io/docs/config_entries_config_flow_handler

mvn23 replied:
The integration fully supports config flows and entries, this schema is a leftover from before that. At the moment it's in place to support upgrades from *really* old versions, but I agree we should not add to this.

Related question: how long are we supposed to keep supporting these kinds of cases? i.e. when can this be removed entirely?

MartinHjelmare replied:
Our normal deprecation period is 2 release cycles, ie two months.

Martidjen replied:
So should I revert the schema to how it was? Or remove it entirely? It wouldn't be compatible with my changes anyway

MartinHjelmare replied:
Please revert it to the existing schema.

Martidjen replied:
Done :)
Thanks for the feedback and sorry for the mess I'm making, it's my first time committing something for Home Assistant and do have to learn a thing or two apparently

mvn23 commented on homeassistant/components/opentherm_gw/climate.py:
```diff
@@ -258,9 +275,14 @@ async def async_set_temperature(self, **kwargs):
-            self._new_target_temperature = await self._gateway.gateway.set_target_temp(
-                temp
-            )
+            if self.setpoint_ovrd_mode == CONF_OVRD_MODE_TEMPORARY:
+                self._new_target_temperature = (
+                    await self._gateway.gateway.set_target_temp(temp, True)
+                )
+            if self.setpoint_ovrd_mode == CONF_OVRD_MODE_CONSTANT:
+                self._new_target_temperature = (
+                    await self._gateway.gateway.set_target_temp(temp, False)
+                )
```
As I understand it, both options are mutually exclusive. If it doesn't match the first, there's no need to check the second. I would suggest to refactor `CONF_SETPOINT_OVRD_MODE` to a boolean value, then there's no need for any `CONF_OVRD_MODE_*` vars and we can just do
```
self._new_target_temperature = (
    await self._gateway.gateway.set_target_temp(temp, self.setpoint_ovrd_mode)
)
```

Martidjen replied:
What you are saying is correct. I was planning on doing a boolean at first, but changed my mind because I thought this would be easier to understand for users from the options panel. If you really suggest to use a boolean I will revert the changes for this PR and add the boolean option in a separate PR (splitting the bullets as Martin said)

mvn23 replied:
It should be safe to assume that users of the gateway are familiar with the terms `temporary` and/or `constant` in relation to temperature overrides (otmonitor uses only these terms). Having a checkbox with one of these terms in the options panel should be clear enough, especially when it is further explained in the docs.
Another option would be to keep the dropdown in the options panel and deriving a boolean from its value upon submission of the form, but in my opinion the checkbox would be sufficient.

Martidjen replied:
Suppose you're right on this one. I will change the option to a boolean, will try to do so in a separate PR.
Would mean I have to update documentation and tests as well but that should be fine.

What do you reckon on the schema in __init__.py, should it be removed now? Or should I just remove the parts I modified?

mvn23 commented on homeassistant/components/opentherm_gw/config_flow.py:
```diff
@@ -121,14 +128,29 @@ async def async_step_init(self, user_input=None):
                             [0, PRECISION_TENTHS, PRECISION_HALVES, PRECISION_WHOLE]
                         ),
                     ),
+                    vol.Optional(
+                        CONF_SETPOINT_OVRD_MODE,
+                        default=self.config_entry.options.get(
+                            CONF_SETPOINT_OVRD_MODE, CONF_OVRD_MODE_TEMPORARY
+                        ),
+                    ): vol.In([CONF_OVRD_MODE_TEMPORARY, CONF_OVRD_MODE_CONSTANT]),
```
See above, I would suggest refactoring to a boolean.

MartinHjelmare commented:
I recommend splitting this PR into two PRs, one per bullet in the description.

Martidjen committed:
Splitting PR

Splitting up the pull request. Setpoint override mode will be added in a separate pull request as a boolean

Martidjen committed:
Remove some leftover includes

Forgot to remove these includes

Martidjen committed:
Fixing strings.json

Martidjen committed:
Fixing sorting of includes

Apparently the includes were supposed to be sorted on one line

Martidjen committed:
Revert schema for yaml config

Changes to the schema for yaml config are not allowed to be changed any more

Martidjen committed:
Re-adding constant CONF_PRECISION

The CONF_PRECISION constant is needed again after reverting the schema for yaml config options

mvn23 reviewed (approved):
Code looks good to me :+1:

MartinHjelmare commented on homeassistant/components/opentherm_gw/climate.py:
```diff
@@ -178,8 +186,8 @@ def unique_id(self):
         """Return the precision of the system."""
-        if self.temp_precision is not None and self.temp_precision != 0:
-            return self.temp_precision
+        if self.temp_read_precision is not None and self.temp_read_precision != 0:
```
```suggestion
        if self.temp_read_precision:
```

MartinHjelmare commented on homeassistant/components/opentherm_gw/climate.py:
```diff
@@ -234,7 +242,11 @@ def target_temperature(self):
     def target_temperature_step(self):
         """Return the supported step of target temperature."""
-        return self.precision
+        if self.temp_set_precision is not None and self.temp_set_precision != 0:
```
```suggestion
        if self.temp_set_precision:
```

MartinHjelmare commented on tests/components/opentherm_gw/test_config_flow.py:
```diff
@@ -189,21 +196,45 @@ async def test_options_form(hass):
 
     result = await hass.config_entries.options.async_configure(
```
Make sure to set up the integration before starting the options flow in the beginning of the test. Use `hass.config_entries.async_setup`. This is needed to make sure that the config flow module is loaded if the test is run in isolation.

Martidjen replied:
Should there also be `with patch() ` in this test form? Or is it enough to place `hass.config_entries.async_setup(entry.entry_id)` and `await hass.async_block_till_done()` under `entry.add_to_hass(hass)`?

MartinHjelmare replied:
We need to patch something so that there's no I/O in the test. If we only care about the isolated options flow, we could patch the integration setup function `async_setup_entry`.

MartinHjelmare replied:
Yes, exactly!

Martidjen replied:
Is it correct to end te test with 
```
assert len(mock_setup.mock_calls) == 0
assert len(mock_setup_entry.mock_calls) == 0
```
Local test will fail if I assert 1
Am I doing something wrong or is this expected?

Martidjen replied:
So might just go without patch.

If I place
```
hass.config_entries.async_setup(entry.entry_id)
await hass.async_block_till_done()
```
and then run the test I only get this `RuntimeWarning: coroutine 'ConfigEntries.async_setup' was never awaited`
Is it ok to leave it at that? Or remove `await hass.async_block_till_done()` as well?

mvn23 replied:
We don't want to go for "as little warnings as possible", but rather for "no warnings whatsoever" :wink: 
Try `await hass.config_entries.async_setup(entry.entry_id)` instead.

Martidjen replied:
If i do that i'm getting `TypeError: async_block_till_done() takes 1 positional argument but 2 were given`

mvn23 replied:
Then you'll have a call to `hass.async_block_till_done()` with an argument somewhere, the test log will probably tell you where.
Rather than troubleshooting error messages one at a time, you can post what you have so far (the full test file, use hastebin or a similar service) and I will see if I can help out with your issues.

Martidjen committed:
Fixing some review comments

Fixing some review comments, still getting a warning:
RuntimeWarning: coroutine 'ConfigEntries.async_setup' was never awaited

Martidjen committed:
Update tests/components/opentherm_gw/test_config_flow.py

As suggested by MartinHjelmare, thank you for your input

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

Martidjen committed:
Fixing indents

Martidjen committed:
Little typo, q should be g

MartinHjelmare commented on homeassistant/components/opentherm_gw/climate.py:
```diff
@@ -79,7 +86,8 @@ def __init__(self, gw_dev, options):
         """Update climate entity options."""
         self.floor_temp = entry.options[CONF_FLOOR_TEMP]
-        self.temp_precision = entry.options[CONF_PRECISION]
+        self.temp_read_precision = entry.options[CONF_READ_PRECISION]
```
Have we released a version of the integration with the `CONF_PRECISION` in the options? Then we need to convert that item to the new options data schema.

Martidjen replied:
CONF_PRECISION was the name of the option before I divided it into read and set precision. CONF_PRECISION was configured in the options flow schema. But I replaced it with CONF_READ_PRECISION. CONF_PRECISION only still exists because it was also in the yaml config schema and I had to revert that to it's original state.

MartinHjelmare replied:
Ok. How should we migrate existing users that had set `CONF_PRECISION`?

Martidjen replied:
Hmm.. good point. I like how well you're thinking this through. 

What I could do is set the const `CONF_READ_PRECISION = read_precision` back to `CONF_READ_PRECISION = precision` so it will use its original setting. Then I could make `CONF_SET_PRECISION` mirror `CONF_READ_PRECISION` when it is not set. 

Would you be ok with that?

MartinHjelmare replied:
I think we can keep the new term named `"read_precision"` since it's more clear what that is now. We can just add a check for the old term too, `"precision"` if `"read_precision"` is not set in entry.options, and use the `"precision"` value for read precision in that case.

Sounds good to default the set precision to the same as the read precision.

Martidjen replied:
After some rethinking I am a bit in doubt. The original `"precision"` setting won't be available from the options form anymore, so when the new settings aren't used it won't be possible to get back to the default setting if `"precision"` was set before. Which somehow forces you to use the new settings.

Would it be possible to set `"read_precision"` to the value of `"precision"` and then set `"precision"` to None? So you won't be bothered by the 'old' setting anymore.

Something like this in the `__init__` or `update_options` functions, or in another function if that would be better?
```
if entry.options[CONF_PRECISION]:
    entry.options[CONF_READ_PRECISION] = entry.options[CONF_PRECISION] 
    entry.options[CONF_PRECISION] = None
self.temp_read_precision = entry.options[CONF_READ_PRECISION]
```
Is something like this possible at all? Or would it be better to use `"precision"` instead of `"read_precision"` after all?
Or am I overthinking stuff right now?

Martidjen committed:
Added migration from old settings to new settings

New settings for precision will both be set to the value of the old precision setting when coming from the previous version of this integration

mvn23 reviewed (changes requested):
See comments. We should also consider adding a test case for the options migration.

Martidjen commented on homeassistant/components/opentherm_gw/__init__.py:
```diff
@@ -94,6 +96,13 @@ async def async_setup_entry(hass, config_entry):
     hass.data[DATA_OPENTHERM_GW][DATA_GATEWAYS][config_entry.data[CONF_ID]] = gateway
 
+    if config_entry.options[CONF_PRECISION]:
+        migrate_options = {}
+        migrate_options[CONF_READ_PRECISION] = config_entry.options[CONF_PRECISION]
+        migrate_options[CONF_SET_PRECISION] = config_entry.options[CONF_PRECISION]
+        migrate_options[CONF_FLOOR_TEMP] = config_entry.options[CONF_FLOOR_TEMP]
```
I see what you mean. But in your example `migrate_options` will be `Null` because `.update()` does not return anything
So it will have to be either:
```
migrate_options = dict(config_entry.options)
migrate_options.update(
    {
        CONF_READ_PRECISION: config_entry.options[CONF_PRECISION],
        CONF_SET_PRECISION: config_entry.options[CONF_PRECISION],
    }
)
del migrate_options[CONF_PRECISION]
```

or something in between like:
```
migrate_options = dict(config_entry.options)
migrate_options[CONF_READ_PRECISION] = config_entry.options[CONF_PRECISION]
migrate_options[CONF_SET_PRECISION] = config_entry.options[CONF_PRECISION]
del migrate_options[CONF_PRECISION]
```

What whould you prefer?

mvn23 replied:
My bad, that's what I get for typing suggestions without testing...
To me the first solution looks cleaner, but others may disagree.

Martidjen replied:
No problem, totally understandable. I agree, will go for the first one then.
Now only have to find a way to test the migration :-)

Martidjen committed:
Fixing check failure and working on review comments

Martidjen committed:
Adding test for options migration

Added test for the options migration when updating from previous version

MartinHjelmare reviewed (approved):
Looks good!

MartinHjelmare commented:
@mvn23 ok with you?

mvn23 reviewed (approved):
Looks good to me :+1:
Thanks @Martidjen for the PR and @MartinHjelmare for the review!
